[Doc Rework] - implementation notes PR1 (#4176)
telefonicaid/fiware-orion: doc/manuals/orion-api.md

If `attrsFormat` is `values` then values partial entity representation mode is used:

```json
{
  "subscriptionId": "12345",
  "data": [ [23, 70], [24] ]
}
```

If `attrsFormat` is `legacy` then subscription representation follows  NGSIv1 format. This way, users 
can benefit from the enhancements of NGSIv2 subscriptions (e.g. filtering) with NGSIv1 legacy notification receivers.

Note that NGSIv1 is deprecated. Thus, we don't recommend to use `legacy` notification format any longer.

```json
{
	"subscriptionId": "56e2ad4e8001ff5e0a5260ec",
	"originator": "localhost",
	"contextResponses": [{
		"contextElement": {
			"type": "Car",
			"isPattern": "false",
			"id": "Car1",
			"attributes": [{
				"name": "temperature",
				"type": "centigrade",
				"value": "26.5",
				"metadatas": [{
					"name": "TimeInstant",
					"type": "recvTime",
					"value": "2015-12-12 11:11:11.123"
				}]
			}]
		},
		"statusCode": {
			"code": "200",
			"reasonPhrase": "OK"
		}
	}]
}
```

> Notifications must include the `Ngsiv2-AttrsFormat` (when using `attrsFormat` = `legacy` is excluded ) 

Reviewer: ```suggestion
Notifications must include the `Ngsiv2-AttrsFormat` (expect when `attrsFormat` is `legacy`) 
```

|----------|------------------------------------------------------------------------------------------|
| `count`  | When used, the total number of types is returned in the HTTP header `Fiware-Total-Count` |
| `values` | When used, the response payload is a JSON array with a list of entity types              |

> | `noAttrDetail` | When used, the request does not provide attribute type details. `types` list associated to each attribute name is set to `[]`. Using this option, Orion solves these queries much faster (up to 30 seconds). |

Reviewer: ```suggestion
| `noAttrDetail` | When used, the request does not provide attribute type details. `types` list associated to each attribute name is set to `[]`. Using this option, Orion solves these queries much faster (in some cases saving from 30 seconds to 0.5 seconds). |
```

|----------|------------------------------------------------------------------------------------------|
| `count`  | When used, the total number of types is returned in the HTTP header `Fiware-Total-Count` |
| `values` | When used, the response payload is a JSON array with a list of entity types              |
| `noAttrDetail` | When used, the request does not provide attribute type details. `types` list associated to each attribute name is set to `[]`. Using this option, Orion solves these queries much faster (in some cases saving from 30 seconds to 0.5 seconds). |

> | `skipForwarding` | when used, CB skips forwarding to CPrs. The query is evaluated using exclusively CB local context information. |

Reviewer: Doesn't apply to GET /v2/types

```suggestion
```

_**Request query parameters**_

| Parameter    | Optional | Type   | Description         | Example |
|--------------|----------|--------|---------------------|---------|
| `entityType` |          | string | Entity Type.        | `Room`  |

_**Request headers**_

| Header               | Optional | Description                                                                                    | Example            |
|----------------------|----------|------------------------------------------------------------------------------------------------|--------------------|
| `Fiware-Service`     | ✓        | Tenant or service. See subsection [Multitency](#multitenancy) for more information.            | `acme`             |
| `Fiware-ServicePath` | ✓        | Service path or subservice. See subsection [Service Path](#service-path) for more information. | `/project`         |

> | `noAttrDetail` | When used, the request does not provide attribute type details. `types` list associated to each attribute name is set to `[]`. Using this option, Orion solves these queries much faster (in some cases saving from 30 seconds to 0.5 seconds). |

Reviewer: `noAttrDetail`, is an `option`, not a header.

```suggestion
```

_**Request query parameters**_

| Parameter    | Optional | Type   | Description         | Example |
|--------------|----------|--------|---------------------|---------|

> | `options` | ✓        | string | Options dictionary.                        | `count` |
> The values that `options` parameter can have for this specific request are:

Reviewer: ```suggestion
| `options` | ✓        | string | Options dictionary.                        | `noAttrDetail` |

The values that `options` parameter can have for this specific request are:
```

_**Request query parameters**_

| Parameter    | Optional | Type   | Description         | Example |
|--------------|----------|--------|---------------------|---------|

> | `entityType` |          | string | Entity Type.        | `Room`  |
> 

Reviewer: Note that entityType is not actually a query parameter, but an URL parameter, e.g.: `GET /v2/types/T1?options=noAttrDetail`)

```suggestion
| `options` | ✓        | string | Options dictionary.                        | `count` |
```

_**Request query parameters**_

| Parameter    | Optional | Type   | Description         | Example |
|--------------|----------|--------|---------------------|---------|
| `entityType` |          | string | Entity Type.        | `Room`  |

> | `options`    | ✓        | string | Options dictionary. | `count` |
> 
> The values that `options` parameter can have for this specific request are:
> 
> | Options  | Description                                                                              |
> |----------|------------------------------------------------------------------------------------------|
> | `skipForwarding` | when used, CB skips forwarding to CPrs. The query is evaluated using exclusively CB local context information. |

Reviewer: Doesn't apply to GET /v2/types

```suggestion
```

| `metadata`         | ✓                 | string  | List of metadata to be included in notification messages. See [Filtering out attributes and metadata](#filtering-out-attributes-and-metadata) section for more detail.            |
| `onlyChangedAttrs` | ✓                 | boolean | List of metadata to be included in notification messages. See [Filtering out attributes and metadata](#filtering-out-attributes-and-metadata) section for more detail.            |
| `timesSent`        | Only on retrieval | number  | Not editable, only present in GET operations. Number of notifications sent due to this subscription.                                                                              |
| `lastNotification` | Only on retrieval | ISO8601 | Not editable, only present in GET operations. Last notification timestamp in ISO8601 format.                                                                                      |
| `lastFailure`      | Only on retrieval | ISO8601 | Not editable, only present in GET operations. Last failure timestamp in ISO8601 format. Not present if subscription has never had a problem with notifications.                   |

> | `lastSuccess`      | Only on retrieval | ISO8601 | Not editable, only present in GET operations. Timestamp in ISO8601 format for last successful notification.  Not present if subscription has never had a successful notification. |

Author: Fixed in https://github.com/telefonicaid/fiware-orion/pull/4176/commits/4c6dd9bc81215b0211286763c8aa4037c38b058b

| [`notification`](#subscriptionnotification) |          | object | An object that describes the notification to send when the subscription is triggered.         |
| `expires`      | ✓        | ISO8601 | Subscription expiration date in ISO8601 format. Permanent subscriptions must omit this field. |
| `status`       |          | string | Either `active` (for active subscriptions) or `inactive` (for inactive subscriptions). If this field is not provided at subscription creation time, new subscriptions are created with the `active` status, which can be changed by clients afterwards. For expired subscriptions, this attribute is set to `expired` (no matter if the client updates it to `active`/`inactive`). Also, for subscriptions experiencing problems with notifications, the status is set to `failed`. As soon as the notifications start working again, the status is changed back to `active`. Additionaly, `oneshot` value is available, firing the notification only once whenever the entity is updated after creating the subscription. Once a notification is triggered, the subscription transitions to "status": "inactive". |

> | `throttling`   | ✓        | number | Minimal period of time in seconds which must elapse between two consecutive notifications. Orion implements this is discarding notifications during the throttling guard period. Thus, notifications may be lost if they arrive too close in time. |

Reviewer: ```suggestion
| `throttling`   | ✓        | number | Minimal period of time in seconds which must elapse between two consecutive notifications. Orion implements this discarding notifications during the throttling guard period. Thus, notifications may be lost if they arrive too close in time. |
```

| `lastSuccessCode`  | Only on retrieval | number  | Not editable, only present in GET operations. the HTTP code (200, 400, 404, 500, etc.) returned by receiving endpoint last time a successful notification was sent (i.e. the success occurred at `lastSuccess` time). Not included in MQTT subscriptions.|
| `failsCounter`     | Only on retrieval | number  | Not editable, only present in GET operations. The number of consecutive failing notifications associated to the subscription. `failsCounter` is increased by one each time a notification attempt fails and reset to 0 if a notification attempt successes (`failsCounter` is ommitted in this case).|
| `maxFailsLimit`    | ✓                 | number  | Establishes a maximum allowed number of consecutive fails. If the number of fails overpasses the value of `maxFailsLimit` (i.e. at a given moment `failsCounter` is greater than `maxFailsLimit`) then Orion automatically passes the subscription to `inactive` state. A subscripiton update operation (`PATCH /v2/subscription/subId`) is needed to re-enable the subscription (setting its state `active` again). |

> Regarding `covered` field, as an example, if `attrs` is `[A, B, C]` for a given subscription, the default behavior 

Reviewer: ```suggestion
Regarding `onlyChangedAttrs` field, as an example, if `attrs` is `[A, B, C]` for a given subscription, the default behavior 
```

| `qs`      | ✓        | object | A key-map of URL query parameters that are included in notification messages.                 |
| `method`  | ✓        | string | The method to use when sending the notification (default is POST). Only valid HTTP methods are allowed. On specifying an invalid HTTP method, a 400 Bad Request error is returned.|
| `payload` | ✓        | string | The payload to be used in notifications. If omitted, the default payload (see [Notification Messages](#notification-messages) sections) is used.|

> | `timeout` | ✓        | number | Maximum time the subscription waits for the response when using in milliseconds. The maximum value allowed for this parameter is 1800000 (30 minutes). If `timeout` is defined to 0 or omitted, then the value passed as `-httpTimeout` CLI parameter is used. See section in the
> [Command line options](admin/cli.md#command-line-options) for more details. |

Reviewer: ```suggestion
| `timeout` | ✓        | number | Maximum time (in milliseconds) the subscription waits for the response. The maximum value allowed for this parameter is 1800000 (30 minutes). If `timeout` is defined to 0 or omitted, then the value passed as `-httpTimeout` CLI parameter is used. See section in the [Command line options](admin/cli.md#command-line-options) for more details. |
```
